### ISSUE TITLE
crashing kernel when trying to load multiple groups with .to_dataset_dict()

### ISSUE THREAD

**Issue by bolliger32** [intake/intake-esm#162]
### Description

Trying to load datasets on the Cheyenne platform. Seemingly when I load multiple groups with to_dataset_dict(), I crash the kernel (if local) or kill workers (if using dask).

### What I Did

Using the Cheyenne Jupyterhub platform, the following works (where `col` is the full collection of cmip6 data on GLADE:
```
dict1 = col.search(source_id='CESM2',variable_id='pr', experiment_id='historical', table_id='Amon').to_dataset_dict()
dict2 = col.search(source_id='CESM2',variable_id='pr', experiment_id='ssp370', table_id='Amon').to_dataset_dict()
```

However, the following breaks, killing the kernel if running locally or killing workers if using dask:
```
dict_tot = col.search(source_id='CESM2',variable_id='pr', experiment_id=['historical','ssp370'], table_id='Amon').to_dataset_dict()
```

### Output of ``intake_esm.__version__``

'2019.10.15.post14'

### Conda list output
```# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
absl-py                   0.8.1                    py37_0    conda-forge
affine                    2.3.0                      py_0    conda-forge
aiohttp                   3.6.1            py37h516909a_0    conda-forge
altair                    3.2.0                    py37_0    conda-forge
antlr-python-runtime      4.7.2                 py37_1000    conda-forge
appdirs                   1.4.3                      py_1    conda-forge
arrow-cpp                 0.15.0           py37h5ac5442_0    conda-forge
asciitree                 0.3.3                      py_2    conda-forge
asn1crypto                1.2.0                    py37_0    conda-forge
astor                     0.7.1                      py_0    conda-forge
async-timeout             3.0.1                   py_1000    conda-forge
atomicwrites              1.3.0                      py_0    conda-forge
attrdict                  2.0.1                      py_0    conda-forge
attrs                     19.3.0                     py_0    conda-forge
backcall                  0.1.0                      py_0    conda-forge
basemap                   1.2.1            py37hd759880_1    conda-forge
beautifulsoup4            4.8.1                    py37_0    conda-forge
binutils_impl_linux-64    2.31.1               h7fc9f1b_2    conda-forge
binutils_linux-64         2.31.1              h6176602_12    conda-forge
bleach                    3.1.0                      py_0    conda-forge
blosc                     1.17.0               he1b5a44_1    conda-forge
bokeh                     1.3.4                    py37_0    conda-forge
boost-cpp                 1.70.0               h8e57a91_2    conda-forge
boto3                     1.9.252                    py_0    conda-forge
botocore                  1.12.252                   py_0    conda-forge
bottleneck                1.2.1           py37h3010b51_1001    conda-forge
branca                    0.3.1                      py_0    conda-forge
brotli                    1.0.7             he1b5a44_1000    conda-forge
bzip2                     1.0.8                h516909a_1    conda-forge
c-ares                    1.15.0            h516909a_1001    conda-forge
ca-certificates           2019.9.11            hecc5488_0    conda-forge
cairo                     1.16.0            hfb77d84_1002    conda-forge
cartopy                   0.17.0          py37hd759880_1006    conda-forge
certifi                   2019.9.11                py37_0    conda-forge
cf-units                  2.1.3            py37hc1659b7_0    conda-forge
cf_units                  2.0.1           py37h3010b51_1002    conda-forge
cffi                      1.13.0           py37h8022711_0    conda-forge
cfgrib                    0.9.7.2                    py_0    conda-forge
cfitsio                   3.470                hb60a0a2_2    conda-forge
cftime                    1.0.3.4         py37hd352d35_1001    conda-forge
chardet                   3.0.4                 py37_1003    conda-forge
click                     7.0                        py_0    conda-forge
click-plugins             1.1.1                      py_0    conda-forge
cligj                     0.5.0                      py_0    conda-forge
climlab                   0.7.5            py37h7eb8c7e_1    conda-forge
climpred                  1.1.0                      py_1    conda-forge
cloudpickle               1.2.2                      py_0    conda-forge
cmocean                   2.0                        py_1    conda-forge
colorama                  0.4.1                      py_0    conda-forge
colorcet                  2.0.1                      py_0    conda-forge
colorspacious             1.1.2              pyh24bf2e0_0    conda-forge
configobj                 5.0.6                      py_0    conda-forge
cryptography              2.7              py37h72c5cf5_0    conda-forge
curl                      7.65.3               hf8cf82a_0    conda-forge
cycler                    0.10.0                     py_1    conda-forge
cython                    0.29.13          py37he1b5a44_0    conda-forge
cytoolz                   0.10.0           py37h516909a_0    conda-forge
dask                      2.6.0                      py_0    conda-forge
dask-core                 2.6.0                      py_0    conda-forge
dask-glm                  0.2.0                      py_1    conda-forge
dask-jobqueue             0.7.0                      py_0    conda-forge
dask-labextension         1.0.0                    pypi_0    pypi
dask-ml                   1.0.0                      py_1    conda-forge
dask-mpi                  2.0.0                    py37_0    conda-forge
datashader                0.8.0                      py_0    conda-forge
datashape                 0.5.4                      py_1    conda-forge
dbus                      1.13.6               he372182_0    conda-forge
decorator                 4.4.0                      py_0    conda-forge
defusedxml                0.6.0                      py_0    conda-forge
descartes                 1.1.0                      py_4    conda-forge
distributed               2.6.0                      py_0    conda-forge
docopt                    0.6.2                      py_1    conda-forge
docrep                    0.2.7                      py_0    conda-forge
docutils                  0.15.2                   py37_0    conda-forge
double-conversion         3.1.5                he1b5a44_1    conda-forge
eccodes                   2.14.1               hec26673_0    conda-forge
entrypoints               0.3                   py37_1000    conda-forge
eofs                      1.4.0                      py_0    conda-forge
erddapy                   0.5.0                      py_0    conda-forge
esmf                      7.1.0             h9a7cb89_1007    conda-forge
esmlab                    2019.4.27                  py_0    conda-forge
esmpy                     7.1.0            py37h24bf2e0_3    conda-forge
expat                     2.2.5             he1b5a44_1003    conda-forge
fastcache                 1.1.0            py37h516909a_0    conda-forge
fasteners                 0.14.1                     py_3    conda-forge
fastparquet               0.3.2            py37hc1659b7_0    conda-forge
fiona                     1.8.6            py37hf242f0b_3    conda-forge
folium                    0.10.0                     py_0    conda-forge
fontconfig                2.13.1            h86ecdb6_1001    conda-forge
freetype                  2.10.0               he983fc9_1    conda-forge
freexl                    1.0.5             h14c3975_1002    conda-forge
fribidi                   1.0.5             h516909a_1002    conda-forge
fsspec                    0.5.2                      py_0    conda-forge
funcargparse              0.2.2                      py_0    conda-forge
future                    0.18.1                   py37_0    conda-forge
gast                      0.3.2                      py_0    conda-forge
gcc_impl_linux-64         7.3.0                habb00fd_2    conda-forge
gcc_linux-64              7.3.0               h553295d_12    conda-forge
gdal                      2.4.1            py37h5f563d9_8    conda-forge
geopandas                 0.6.1                      py_0    conda-forge
geos                      3.7.2                he1b5a44_2    conda-forge
geotiff                   1.4.3             hb6868eb_1001    conda-forge
geoviews                  1.6.5                      py_0    conda-forge
geoviews-core             1.6.5                      py_0    conda-forge
gettext                   0.19.8.1          hc5be6a0_1002    conda-forge
gflags                    2.2.2             he1b5a44_1001    conda-forge
giflib                    5.1.7                h516909a_1    conda-forge
gitdb2                    2.0.6                      py_0    conda-forge
gitpython                 3.0.3                      py_0    conda-forge
glib                      2.58.3            h6f030ca_1002    conda-forge
glog                      0.4.0                he1b5a44_1    conda-forge
gmp                       6.1.2             hf484d3e_1000    conda-forge
gmpy2                     2.1.0b1          py37h04dde30_0    conda-forge
graphite2                 1.3.13            hf484d3e_1000    conda-forge
graphviz                  2.40.1               h0f2764d_1    conda-forge
grpc-cpp                  1.23.0               h18db393_0    conda-forge
grpcio                    1.23.0           py37he9ae1f9_0    conda-forge
gsl                       2.5                  h294904e_1    conda-forge
gst-plugins-base          1.14.5               h0935bb2_0    conda-forge
gstreamer                 1.14.5               h36ae1b5_0    conda-forge
gsw                       3.3.1            py37h516909a_0    conda-forge
gxx_impl_linux-64         7.3.0                hdf63c60_2    conda-forge
gxx_linux-64              7.3.0               h553295d_12    conda-forge
h5netcdf                  0.7.4                      py_0    conda-forge
h5py                      2.10.0          nompi_py37h513d04c_100    conda-forge
harfbuzz                  2.4.0                h9f30f68_3    conda-forge
hdf4                      4.2.13            h9a582f1_1002    conda-forge
hdf5                      1.10.5          nompi_h3c11f04_1103    conda-forge
heapdict                  1.0.1                      py_0    conda-forge
holoviews                 1.12.6                     py_0    conda-forge
hvplot                    0.5.1                      py_0    conda-forge
icu                       64.2                 he1b5a44_1    conda-forge
idna                      2.8                   py37_1000    conda-forge
imageio                   2.6.1                    py37_0    conda-forge
importlib_metadata        0.23                     py37_0    conda-forge
intake                    0.5.3                      py_0    conda-forge
intake-esm                2019.10.15.post14          pypi_0    pypi
intake-stac               0.2.0                      py_0    conda-forge
intake-xarray             0.3.1                      py_0    conda-forge
ipykernel                 5.1.2            py37h5ca1d4c_0    conda-forge
ipyleaflet                0.11.4                   py37_1    conda-forge
ipython                   7.8.0            py37h5ca1d4c_0    conda-forge
ipython_genutils          0.2.0                      py_1    conda-forge
ipywidgets                7.5.1                      py_0    conda-forge
iris                      2.2.0                 py37_1003    conda-forge
jasper                    1.900.1           h07fcdf6_1006    conda-forge
jedi                      0.15.1                   py37_0    conda-forge
jinja2                    2.10.3                     py_0    conda-forge
jmespath                  0.9.4                      py_0    conda-forge
joblib                    0.14.0                     py_0    conda-forge
jpeg                      9c                h14c3975_1001    conda-forge
json-c                    0.13.1            h14c3975_1001    conda-forge
json5                     0.8.5                      py_0    conda-forge
jsonschema                3.1.1                    py37_0    conda-forge
jupyter                   1.0.0                      py_2    conda-forge
jupyter-server-proxy      1.1.0                      py_0    conda-forge
jupyter_client            5.3.3                    py37_1    conda-forge
jupyter_console           6.0.0                      py_0    conda-forge
jupyter_contrib_core      0.3.3                      py_2    conda-forge
jupyter_contrib_nbextensions 0.5.1                    py37_0    conda-forge
jupyter_core              4.5.0                      py_0    conda-forge
jupyter_highlight_selected_word 0.2.0                 py37_1000    conda-forge
jupyter_latex_envs        1.4.4                 py37_1000    conda-forge
jupyter_nbextensions_configurator 0.4.1                    py37_0    conda-forge
jupyterlab                1.1.4                      py_0    conda-forge
jupyterlab_server         1.0.6                      py_0    conda-forge
kealib                    1.4.10            h58c409b_1005    conda-forge
keras                     2.3.1                    py37_0    conda-forge
keras-applications        1.0.8                      py_1    conda-forge
keras-preprocessing       1.1.0                      py_0    conda-forge
kiwisolver                1.1.0            py37hc9558a2_0    conda-forge
krb5                      1.16.3            h05b26f9_1001    conda-forge
libaec                    1.0.4                hf484d3e_0    conda-forge
libblas                   3.8.0               14_openblas    conda-forge
libcblas                  3.8.0               14_openblas    conda-forge
libclang                  9.0.0                hc9558a2_1    conda-forge
libcurl                   7.65.3               hda55be3_0    conda-forge
libdap4                   3.20.4               hd3bb157_0    conda-forge
libedit                   3.1.20170329      hf8c457e_1001    conda-forge
libevent                  2.1.10               h72c5cf5_0    conda-forge
libffi                    3.2.1             he1b5a44_1006    conda-forge
libgcc-ng                 9.1.0                hdf63c60_0  
libgdal                   2.4.1                heae24aa_8    conda-forge
libgfortran-ng            7.3.0                hdf63c60_2    conda-forge
libgpuarray               0.7.6             h14c3975_1003    conda-forge
libiconv                  1.15              h516909a_1005    conda-forge
libkml                    1.3.0             h4fcabce_1010    conda-forge
liblapack                 3.8.0               14_openblas    conda-forge
libllvm9                  9.0.0                hc9558a2_2    conda-forge
libnetcdf                 4.6.2             h303dfb8_1003    conda-forge
libopenblas               0.3.7                h6e990d7_2    conda-forge
libpng                    1.6.37               hed695b0_0    conda-forge
libpq                     11.5                 hd9ab2ff_1    conda-forge
libprotobuf               3.8.0                h8b12597_0    conda-forge
libsodium                 1.0.17               h516909a_0    conda-forge
libspatialindex           1.9.0                he1b5a44_1    conda-forge
libspatialite             4.3.0a            h79dc798_1030    conda-forge
libssh2                   1.8.2                h22169c7_2    conda-forge
libstdcxx-ng              9.1.0                hdf63c60_0  
libtiff                   4.0.10            h57b8799_1003    conda-forge
libtool                   2.4.6             h14c3975_1002    conda-forge
libuuid                   2.32.1            h14c3975_1000    conda-forge
libuv                     1.32.0               h516909a_0    conda-forge
libxcb                    1.13              h14c3975_1002    conda-forge
libxkbcommon              0.8.4                h516909a_0    conda-forge
libxml2                   2.9.9                hee79883_5    conda-forge
libxslt                   1.1.33               h31b3aaa_0    conda-forge
line_profiler             2.1.2           py37h516909a_1003    conda-forge
llvmlite                  0.29.0           py37hfd453ef_1    conda-forge
locket                    0.2.0                      py_2    conda-forge
lxml                      4.4.1            py37h7ec2d77_0    conda-forge
lz4                       2.2.1            py37hd79334b_0    conda-forge
lz4-c                     1.8.3             he1b5a44_1001    conda-forge
mako                      1.1.0                      py_0    conda-forge
markdown                  3.1.1                      py_0    conda-forge
markupsafe                1.1.1            py37h14c3975_0    conda-forge
matplotlib                3.1.1                    py37_1    conda-forge
matplotlib-base           3.1.1            py37he7580a8_1    conda-forge
mechanicalsoup            0.12.0                     py_0    conda-forge
memory_profiler           0.55.0                     py_0    conda-forge
mercantile                1.1.2                      py_0    conda-forge
metpy                     0.11.0                   py37_1    conda-forge
mistune                   0.8.4           py37h14c3975_1000    conda-forge
mock                      3.0.5                    py37_0    conda-forge
monotonic                 1.5                        py_0    conda-forge
more-itertools            7.2.0                      py_0    conda-forge
mpc                       1.1.0             hb20f59a_1006    conda-forge
mpfr                      4.0.2                he80fd80_0    conda-forge
mpi                       1.0                       mpich    conda-forge
mpi4py                    3.0.2            py37hcf07815_0    conda-forge
mpich                     3.2.1             hc99cbb1_1014    conda-forge
mpmath                    1.1.0                      py_0    conda-forge
msgpack-numpy             0.4.4.3                    py_0    conda-forge
msgpack-python            0.6.2            py37hc9558a2_0    conda-forge
multidict                 4.5.2           py37h14c3975_1000    conda-forge
multipledispatch          0.6.0                      py_0    conda-forge
munch                     2.3.2                      py_0    conda-forge
nbconvert                 5.6.0                    py37_1    conda-forge
nbdime                    1.1.0                    py37_0    conda-forge
nbformat                  4.4.0                      py_1    conda-forge
nbgitpuller               0.7.2                      py_0    conda-forge
nc-time-axis              1.2.0                      py_0    conda-forge
ncar-jobqueue             2019.10.16.1             pypi_0    pypi
ncurses                   6.1               hf484d3e_1002    conda-forge
netcdf-fortran            4.4.5             h0789656_1004    conda-forge
netcdf4                   1.5.1.2          py37h73a1b54_1    conda-forge
networkx                  2.4                        py_0    conda-forge
ninja                     1.9.0                h6bb024c_0    conda-forge
nodejs                    12.11.1              h10a4023_0    conda-forge
notebook                  6.0.1                    py37_0    conda-forge
nspr                      4.20              hf484d3e_1000    conda-forge
nss                       3.46                 he751ad9_0    conda-forge
numba                     0.45.1           py37hb3f55d8_0    conda-forge
numcodecs                 0.6.3            py37hf484d3e_0    conda-forge
numpy                     1.17.2           py37h95a1406_0    conda-forge
olefile                   0.46                       py_0    conda-forge
openjpeg                  2.3.1                h21c5421_1    conda-forge
openssl                   1.1.1c               h516909a_0    conda-forge
owslib                    0.18.0                     py_0    conda-forge
packaging                 19.2                       py_0    conda-forge
palettable                3.3.0                      py_0    conda-forge
pandas                    0.25.1           py37hb3f55d8_0    conda-forge
pandoc                    2.7.3                         0    conda-forge
pandocfilters             1.4.2                      py_1    conda-forge
panel                     0.6.4                         0    conda-forge
pango                     1.42.4               ha030887_1    conda-forge
param                     1.9.2                      py_0    conda-forge
parquet-cpp               1.5.1                         2    conda-forge
parso                     0.5.1                      py_0    conda-forge
partd                     1.0.0                      py_0    conda-forge
patsy                     0.5.1                      py_0    conda-forge
pcre                      8.43                 he1b5a44_0    conda-forge
pexpect                   4.7.0                    py37_0    conda-forge
pickleshare               0.7.5                 py37_1000    conda-forge
pillow                    6.2.0            py37h6b7be26_0    conda-forge
pint                      0.9                      py37_2    conda-forge
pip                       19.3                     py37_0    conda-forge
pixman                    0.38.0            h516909a_1003    conda-forge
pluggy                    0.13.0                   py37_0    conda-forge
pooch                     0.5.2                    py37_0    conda-forge
poppler                   0.67.0               ha967d66_7    conda-forge
poppler-data              0.4.9                         1    conda-forge
postgresql                11.5                 hc63931a_1    conda-forge
proj4                     5.2.0             he1b5a44_1006    conda-forge
prometheus_client         0.7.1                      py_0    conda-forge
prompt_toolkit            2.0.10                     py_0    conda-forge
properscoring             0.1                        py_0    conda-forge
protobuf                  3.8.0            py37he1b5a44_1    conda-forge
psutil                    5.6.3            py37h516909a_0    conda-forge
psyplot                   1.2.1                    py37_0    conda-forge
pthread-stubs             0.4               h14c3975_1001    conda-forge
ptyprocess                0.6.0                   py_1001    conda-forge
py                        1.8.0                      py_0    conda-forge
pyarrow                   0.15.0           py37h8b68381_1    conda-forge
pycparser                 2.19                     py37_1    conda-forge
pyct                      0.4.6                      py_0    conda-forge
pyct-core                 0.4.6                      py_0    conda-forge
pydap                     3.2.2                 py37_1000    conda-forge
pyepsg                    0.4.0                      py_0    conda-forge
pygments                  2.4.2                      py_0    conda-forge
pygpu                     0.7.6           py37h3010b51_1000    conda-forge
pyinterp                  0.0.6            py37h97f2665_0    conda-forge
pykdtree                  1.3.1           py37h3010b51_1002    conda-forge
pyke                      1.1.1                 py37_1000    conda-forge
pyopenssl                 19.0.0                   py37_0    conda-forge
pyorbital                 1.5.0                      py_0    conda-forge
pyparsing                 2.4.2                      py_0    conda-forge
pyproj                    1.9.6           py37h516909a_1002    conda-forge
pyqt                      5.12.3           py37hcca6a23_0    conda-forge
pyqt5-sip                 4.19.18                  pypi_0    pypi
pyqtwebengine             5.12.1                   pypi_0    pypi
pyresample                1.13.2           py37hb3f55d8_0    conda-forge
pyrsistent                0.15.4           py37h516909a_0    conda-forge
pyshp                     2.1.0                      py_0    conda-forge
pysocks                   1.7.1                    py37_0    conda-forge
pyspectral                0.9.2                      py_0    conda-forge
pyspharm                  1.0.9           py37h7eb8c7e_1004    conda-forge
pytest                    5.2.1                    py37_0    conda-forge
python                    3.7.3                h33d41f4_1    conda-forge
python-blosc              1.8.1            py37hf484d3e_0    conda-forge
python-dateutil           2.7.5                      py_0    conda-forge
python-geotiepoints       1.1.8            py37hd352d35_0    conda-forge
python-graphviz           0.13                       py_0    conda-forge
python-snappy             0.5.4            py37hee44bf9_0    conda-forge
python-wget               3.2                        py_0    conda-forge
pytide                    1.1.0            py37h9de70de_0    conda-forge
pytorch-cpu               1.1.0            py37he1b5a44_0    conda-forge
pytz                      2019.3                     py_0    conda-forge
pyviz_comms               0.7.2                      py_0    conda-forge
pywavelets                1.0.3            py37hd352d35_1    conda-forge
pyyaml                    5.1.2            py37h516909a_0    conda-forge
pyzmq                     18.1.0           py37h1768529_0    conda-forge
qt                        5.12.5               h0c104cb_0    conda-forge
qtconsole                 4.5.5                      py_0    conda-forge
rasterio                  1.0.25           py37hdff7cfa_0    conda-forge
re2                       2019.09.01           he1b5a44_0    conda-forge
readline                  8.0                  hf8c457e_0    conda-forge
regionmask                0.4.0                      py_0    conda-forge
requests                  2.22.0                   py37_1    conda-forge
rio-cogeo                 1.1.5                      py_0    conda-forge
rtree                     0.8.3           py37h666c49c_1002    conda-forge
ruamel.yaml               0.16.5           py37h516909a_1    conda-forge
ruamel.yaml.clib          0.2.0            py37h516909a_0    conda-forge
s3transfer                0.2.1                    py37_0    conda-forge
sat-stac                  0.3.0                      py_0    conda-forge
satpy                     0.17.1             pyhca3e704_0    conda-forge
scikit-image              0.16.1           py37hb3f55d8_0    conda-forge
scikit-learn              0.21.3           py37hcdab131_0    conda-forge
scipy                     1.3.1            py37h921218d_2    conda-forge
seaborn                   0.9.0                      py_1    conda-forge
seawater                  3.3.4                      py_1    conda-forge
send2trash                1.5.0                      py_0    conda-forge
setuptools                41.4.0                   py37_0    conda-forge
shapely                   1.6.4           py37hec07ddf_1006    conda-forge
simpervisor               0.3                        py_1    conda-forge
siphon                    0.8.0                 py37_1001    conda-forge
six                       1.12.0                py37_1000    conda-forge
smmap2                    2.0.5                      py_0    conda-forge
snakeviz                  2.0.1                      py_0    conda-forge
snappy                    1.1.7             he1b5a44_1002    conda-forge
snuggs                    1.4.7                      py_0    conda-forge
sortedcontainers          2.1.0                      py_0    conda-forge
soupsieve                 1.9.4                    py37_0    conda-forge
sparse                    0.8.0                      py_0    conda-forge
sqlite                    3.30.1               hcee41ef_0    conda-forge
statsmodels               0.10.1           py37hc1659b7_0    conda-forge
supermercado              0.0.5                      py_0    conda-forge
sympy                     1.4                      py37_0    conda-forge
tabulate                  0.8.5                      py_0    conda-forge
tbb                       2018.0.5             h2d50403_0    conda-forge
tblib                     1.4.0                      py_0    conda-forge
tensorboard               1.13.1                   py37_0    conda-forge
tensorflow                1.13.1               h5ece82f_5    conda-forge
tensorflow-base           1.13.1           py37h5ece82f_5    conda-forge
tensorflow-estimator      1.13.0                     py_0  
termcolor                 1.1.0                      py_2    conda-forge
terminado                 0.8.2                    py37_0    conda-forge
testpath                  0.4.2                   py_1001    conda-forge
theano                    1.0.4           py37hf484d3e_1000    conda-forge
thrift                    0.11.0          py37hf484d3e_1001    conda-forge
thrift-cpp                0.12.0            hf3afdfd_1004    conda-forge
tiledb                    1.6.2                h69c774e_1    conda-forge
tiledb-py                 0.4.3            py37hb78b526_1    conda-forge
tk                        8.6.9             hed695b0_1003    conda-forge
toolz                     0.10.0                     py_0    conda-forge
tornado                   6.0.3            py37h516909a_0    conda-forge
tqdm                      4.36.1                     py_0    conda-forge
traitlets                 4.3.3                    py37_0    conda-forge
traittypes                0.2.1                      py_1    conda-forge
trollimage                1.10.1                     py_0    conda-forge
trollsift                 0.3.3                      py_0    conda-forge
tzcode                    2019a             h516909a_1002    conda-forge
udunits2                  2.2.27.6          h4e0c4b3_1001    conda-forge
uriparser                 0.9.3                he1b5a44_1    conda-forge
urllib3                   1.25.6                   py37_0    conda-forge
vega                      2.6.0                    py37_0    conda-forge
vega_datasets             0.7.0                      py_0    conda-forge
vincent                   0.4.4                      py_1    conda-forge
viscm                     0.7                pyh24bf2e0_0    conda-forge
watermark                 2.0.1                      py_0    conda-forge
wcwidth                   0.1.7                      py_1    conda-forge
webencodings              0.5.1                      py_1    conda-forge
webob                     1.8.5                      py_0    conda-forge
werkzeug                  0.16.0                     py_0    conda-forge
wheel                     0.33.6                   py37_0    conda-forge
widgetsnbextension        3.5.1                    py37_0    conda-forge
windspharm                1.7.0                 py37_1001    conda-forge
wrapt                     1.11.2           py37h516909a_0    conda-forge
wrf-python                1.3.2            py37h7eb8c7e_1    conda-forge
xarray                    0.14.0                     py_0    conda-forge
xerces-c                  3.2.2             h8412b87_1004    conda-forge
xesmf                     0.2.1                      py_0    conda-forge
xgcm                      0.2.0                      py_0    conda-forge
xhistogram                0.1.1                      py_0    conda-forge
xlrd                      1.2.0                      py_0    conda-forge
xorg-kbproto              1.0.7             h14c3975_1002    conda-forge
xorg-libice               1.0.10               h516909a_0    conda-forge
xorg-libsm                1.2.3             h84519dc_1000    conda-forge
xorg-libx11               1.6.9                h516909a_0    conda-forge
xorg-libxau               1.0.9                h14c3975_0    conda-forge
xorg-libxdmcp             1.1.3                h516909a_0    conda-forge
xorg-libxext              1.3.4                h516909a_0    conda-forge
xorg-libxpm               3.5.12            h516909a_1002    conda-forge
xorg-libxrender           0.9.10            h516909a_1002    conda-forge
xorg-libxt                1.1.5             h516909a_1003    conda-forge
xorg-renderproto          0.11.1            h14c3975_1002    conda-forge
xorg-xextproto            7.3.0             h14c3975_1002    conda-forge
xorg-xproto               7.0.31            h14c3975_1007    conda-forge
xrft                      0.2.0                      py_0    conda-forge
xskillscore               0.0.7                      py_0    conda-forge
xz                        5.2.4             h14c3975_1001    conda-forge
yaml                      0.1.7             h14c3975_1001    conda-forge
yarl                      1.3.0           py37h14c3975_1000    conda-forge
zarr                      2.3.2                    py37_0    conda-forge
zeromq                    4.3.2                he1b5a44_2    conda-forge
zict                      1.0.0                      py_0    conda-forge
zipp                      0.6.0                      py_0    conda-forge
zlib                      1.2.11            h516909a_1006    conda-forge
zstd                      1.4.0                h3b9ef0a_0    conda-forge```

**Comment by bolliger32** [comment]
OK so this problem occurs when multiple threads are trying to run `intake_esm.merge_util._aggregate`. If you are using the threaded dask scheduler, it breaks. And if you are using distributed with `ncar_jobqueue.NCARCluster` with multiple threads per job (the default), then you again have problems. I wonder if it would make sense to set the default in `NCARCluster` to having nprocs = ncores and have 1 thread per process. Threads just seem to cause all kinds of problems with python. But perhaps there are other reasons to keep the defaults as they are.

**Comment by andersy005** [comment]
@bolliger32,

Thank you for the detailed description of the problem.  What happens if you provide the chunks argument to `to_dataset_dict()`? 

For instance:


```python
dict1 = col.search(source_id='CESM2',variable_id='pr', experiment_id='historical', 

table_id='Amon').to_dataset_dict(cdf_kwargs={"chunks":  {"time" : 36}})
```



> I wonder if it would make sense to set the default in NCARCluster to having nprocs = ncores and have 1 thread per process. 

One of the things I have noticed is that when it comes to netCDF, the I/O performance can be terrible depending on how the dask cluster is set up (multi-threaded workers/processes vs single threaded). 


What's the output of 

```python
print(cluster.job_script())
```

when you ran into this issue?



It's worth running some benchmarks with single threaded workers, for instance

```python
cluster = NCARCluster(processes=4, cores=4, memory="100GB)
cluster.adapt(minimum_jobs=1, maximum_jobs=5)
```

and comparing these with the multi-threaded workers (with `cluster = NCARCluster()`, you get a single process with 36 threads per job on Cheyenne).

I will look into these micro-benchmarks and will report back.

**Comment by bolliger32** [comment]
Thanks for looking into it! The chunks don't change the issue unfortunately. In other words, with the threaded scheduler (the default in jupyterhub when using it on cheyenne) or with the standard specs for NCARCluster on Cheyenne, any call to `to_dataset_dict` that results in >1 group kills the kernel (locally) or the workers (distributed).

This is the outut of `cluster.job_script()`:
```
#!/usr/bin/env bash

#PBS -N dask-worker
#PBS -q regular
#PBS -A UCGD0006
#PBS -l select=1:ncpus=36:mem=109GB
#PBS -l walltime=01:00:00
#PBS -e /glade/scratch/bolliger/
#PBS -o /glade/scratch/bolliger/
JOB_ID=${PBS_JOBID%%.*}

/glade/u/home/bolliger/miniconda3/envs/CMIP6-201910-precip/bin/python -m distributed.cli.dask_worker tcp://10.148.10.21:37671 --nthreads 36 --memory-limit 109.00GB --name name --nanny --death-timeout 60 --local-directory /glade/scratch/bolliger --interface ib0
```

I have since tried to run with 1 thread per process, i.e. by calling `NCARCluster(project='UCGD0006', processes=18, cores=18, resource_spec='select=1:ncpus=18:mem=109GB')`. This doesn't get the immediate killed workers on the call to `to_dataset_dict`, however when I try to `persist()` all of the datasets that I'm working with, I get a ton of killed workers even though it doesn't seem like they should be running out of memory. I've tried increasing the number of workers and I've tried (as indicated above) limiting CPUs to 18 because I'm using the share queue and [this site](https://www2.cisl.ucar.edu/resources/computational-systems/cheyenne/running-jobs/job-submission-queues) seems to say that I cant have more than 18? But In every single one of these cases, I haven't been able to persist the dataset objects. Overall, the netcdfs in my collection take up ~60GB, so they should be able to load on one node, if not 2 (which I have also tried). So I'm not quite sure why they keep killing the workers, but now I'm stuck at that issue.

**Comment by bolliger32** [comment]
Yeah - still haven't been able to figure out how to load this data without killing workers. It seems to be random which datasets cause the workers to be killed each time. And it really doesn't seem like a memory issue. I've now started going one at a time, persisting, `wait`ing, and then reducing the dataset by several `mean` operations (each time persisting and waiting). So by the time I loop to the next dataset in the dictionary, the original one is <1MB. No idea why these workers keep being killed

**Comment by andersy005** [comment]
> I have since tried to run with 1 thread per process, i.e. by calling NCARCluster(project='UCGD0006', processes=36)

With this setup, your 109GB is being divided among the 36 processes --> 1 worker/process has ~ 3GB Of memory. I would recommend running a few number of processes (3 or 4 maybe) giving your workers "enough" memory (~>25GB per worker).

**Comment by bolliger32** [comment]
I have tried that to no avail unfortunately. I have also tried "tricking" the NCARCluster object by giving it processes=36 and memory='400G' b/c the actual memory usage still never exceeds ~30G on the node but this way dask shouldn't kill workers due to memory issues. I don't think it's the memory issue. I frequently get errors that look like communication errors. Happy to send you a short notebook to run through if it would help debug. And/or to try any other thoughts you might have.
